presto-jdbc/src/main/java/io/prestosql/jdbc/PrestoResultSet.java

```diff
@@ -1755,6 +1763,67 @@ private static Number toNumber(Object value)
         return (maxRows > 0) ? new LengthLimitedIterator<>(rowsIterator, maxRows) : rowsIterator;
     }
 
+    private static class AsyncIterator<T>
+            implements Iterator<T>
+    {
+        private static final int MAX_QUEUED_ROWS = 50_000;
+        private static ThreadFactoryBuilder threadFactoryBuilder = new ThreadFactoryBuilder().setDaemon(true)
+                .setNameFormat("async-client-datafetch-%d");
+
+        private static final ExecutorService executorService = Executors.newCachedThreadPool(threadFactoryBuilder.build());
+
+        private final BlockingQueue<T> rowQueue = new LinkedBlockingQueue<>(MAX_QUEUED_ROWS);
+        Iterator<T> dataItr;
+        CompletableFuture<Void> future;
+
+        public AsyncIterator(Iterator<T> dataItr)
+        {
+            this.dataItr = dataItr;
+            future = CompletableFuture.supplyAsync(() -> {
+                while (dataItr.hasNext()) {
+                    try {
+                        rowQueue.put(dataItr.next());
+                    }
+                    catch (InterruptedException ex) {
+                        throw new RuntimeException(ex);
+                    }
+                }
+                return null;
+            }, executorService);
+        }
+
+        public void cancel()
+        {
+            future.cancel(true);
+        }
+
+        @Override
+        public boolean hasNext()
+        {
+            if (!rowQueue.isEmpty()) {
+                return true;
+            }
+            while (rowQueue.isEmpty() && !future.isDone()) {
+                // making sure rowQueue has some records to process oe will return false
+            }
+            if (future.isCompletedExceptionally()) {
+                try {
+                    future.get();
+                }
+                catch (ExecutionException | InterruptedException ex) {
```

Comment:
Similar comment to above about restoring interrupt status and unwrapping `ExecutionException`